### PR TITLE
[SPARK-559] Changes for strict mode CI tests

### PR DESCRIPTION
**Pull request by susanxhuynh**
Changes for the tests to run in strict mode
* Makefile change to spin up a strict cluster
* Setup service account, grant permissions
** Moved from old `setup_permissions.sh` script to sdk_security module
* Set a default "/spark" app name; granted permissions to drivers with that app name
* Marked test_jar() and test_marathon_group(), which runs test_jar(), as “expected fail”; failures tracked by SPARK-593

### CODE REVIEW

**Comment by susanxhuynh** [comment]
cc @ArtRand @skonto

File: Makefile [hunk 1]
```diff
@@ -135,6 +135,7 @@ test-env:
 	source test-env/bin/activate
 	pip3 install -r tests/requirements.txt
 
+CF_TEMPLATE_URL ?= https://s3.amazonaws.com/downloads.mesosphere.io/dcos-enterprise/testing/master/cloudformation/ee.single-master.cloudformation.json
```
Comment (ArtRand) [inline Makefile]: this is the strict CF template?

File: tests/test_kafka.py [hunk 2]
```diff
@@ -140,7 +151,7 @@ def test_pipeline(kerberos_flag, stop_count, jar_uri, keytab_secret, jaas_uri=No
         "--conf", "spark.mesos.driver.secret.filenames=kafka-client.keytab",
         "--conf", "spark.mesos.executor.secret.names={}".format(keytab_secret),
         "--conf", "spark.mesos.executor.secret.filenames=kafka-client.keytab",
-        "--conf", "spark.mesos.task.labels=DCOS_SPACE:/spark",
+        "--conf", "spark.mesos.task.labels=DCOS_SPACE:{}".format(utils.SPARK_APP_NAME),
```
Comment (ArtRand) [inline tests/test_kafka.py]: This is done [in the CLI](https://github.com/mesosphere/spark-build/blob/master/cli/dcos-spark/submit_builder.go#L651) and iff you want to specify a different DCOS_SPACE, it should be done as a [DC/OS arg](https://github.com/mesosphere/spark-build/blob/master/cli/dcos-spark/main.go#L25) - I can make the PR to change this, however.

File: tests/test_spark.py [hunk 3]
```diff
@@ -187,15 +187,15 @@ def test_s3():
     utils.run_tests(app_url=utils._scala_test_jar_url(),
                     app_args=app_args,
                     expected_output="Read 3 lines",
-                    app_name="/spark",
                     args=args)
 
 
+@pytest.mark.xfail(utils.is_strict(), reason="Currently fails in strict mode")
```
Comment (ArtRand) [inline tests/test_spark.py]: We should change this test to something that can be exercised in `strict`, probably still good to test this as folding (marathon groups) is becoming the norm.

File: tests/utils.py [hunk 4]
```diff
@@ -2,18 +2,26 @@
...
 
-import json
 import logging
 import os
 import re
-import requests
 import shakedown
 import subprocess
 import urllib
+import urllib.parse
```
Comment (ArtRand) [inline tests/utils.py]: is there a reason this can't be done with the above import?

File: tests/utils.py [hunk 5]
```diff
@@ -175,11 +179,10 @@ def upload_file(file_path):
     return s3.http_url(basename)
 
 
-def submit_job(app_url, app_args, app_name="/spark", args=[]):
+def submit_job(app_url, app_args, app_name=SPARK_APP_NAME, args=[]):
     if is_strict():
-        args += ["--conf", 'spark.mesos.driverEnv.MESOS_MODULES=file:///opt/mesosphere/etc/mesos-scheduler-modules/dcos_authenticatee_module.json']
-        args += ["--conf", 'spark.mesos.driverEnv.MESOS_AUTHENTICATEE=com_mesosphere_dcos_ClassicRPCAuthenticatee']
-        args += ["--conf", 'spark.mesos.principal=service-acct']
+        args += ["--conf", 'spark.mesos.containerizer=mesos']
```
Comment (ArtRand) [inline tests/utils.py]: is MESOS containerizer required for `strict`? (wasn't in the past)

File: tests/utils.py [hunk 6]
```diff
@@ -241,3 +244,55 @@ def teardown_spark():
 
 def _scala_test_jar_url():
     return s3.http_url(os.path.basename(os.environ["SCALA_TEST_JAR_PATH"]))
+
+
+def spark_security_session():
+    '''
+    Spark strict mode setup is slightly different from dcos-commons, so can't use sdk_security::security_session.
```
Comment (ArtRand) [inline tests/utils.py]: explain why?

File: tests/utils.py [hunk 7]
```diff
@@ -241,3 +244,55 @@ def teardown_spark():
...
+    '''
+    user = 'nobody'
+    role = '*'
+    service_account = SPARK_SERVICE_ACCOUNT
+    secret = SPARK_SERVICE_ACCOUNT_SECRET
+
+    def grant_driver_permission(service_account_name, app_id):
+        dcosurl, headers = sdk_security.get_dcos_credentials()
+        app_id_encoded = urllib.parse.quote(app_id, safe='')
+        app_id_encoded = urllib.parse.quote(app_id_encoded, safe='')  # double-encoded (why?)
```
Comment (ArtRand) [inline tests/utils.py]: could you nest these so there isn't two assignments?

File: tests/utils.py [hunk 8]
```diff
@@ -241,3 +244,55 @@ def teardown_spark():
 
 def _scala_test_jar_url():
     return s3.http_url(os.path.basename(os.environ["SCALA_TEST_JAR_PATH"]))
+
+
+def spark_security_session():
```
Comment (ArtRand) [inline tests/utils.py]: 👍

File: Makefile (hunk 1 above)
Comment (susanxhuynh) [reply Makefile]: This is permissive by default. The CI job in TeamCity (spark-build-PR-strict) has this set to a strict template.

File: tests/test_kafka.py (hunk 2 above)
Comment (susanxhuynh) [reply tests/test_kafka.py]: OK, I'll let you change it in a separate PR.

File: tests/test_spark.py (hunk 3 above)
Comment (susanxhuynh) [reply tests/test_spark.py]: Yeah, guess it doesn't matter much which test it is. I'll change it to SparkPi.

File: tests/utils.py (hunk 4 above)
Comment (susanxhuynh) [reply tests/utils.py]: If I only import `urllib`, the following line will fail:
`app_id_encoded = urllib.parse.quote(app_id, safe='')`

File: tests/utils.py (hunk 5 above)
Comment (susanxhuynh) [reply tests/utils.py]: Let me double-check ... I don't remember why I added it.

File: tests/utils.py (hunk 6 above)
Comment (susanxhuynh) [reply tests/utils.py]: I'll add more details about what's different: (1) the role is "*", (2) the driver permissions to execute tasks.

File: tests/utils.py (hunk 7 above)
Comment (susanxhuynh) [reply tests/utils.py]: Sure

**Comment by susanxhuynh** [comment]
@ArtRand I remember why I added mesos containerizer: it was either that or needing to set SPARK_USER in docker containerizer. I've changed it to the latter, since that's our current default (docker).

**Comment by susanxhuynh** [comment]
@ArtRand Addressed your comments. PTAL.

**Comment by ArtRand** [comment]
LGTM Thanks @susanxhuynh

**Pull request merged**